index.js

```diff
@@ -296,21 +312,64 @@ function createFocusTrap(element, userOptions) {
   // kind of need to capture the action at the keydown phase.
   function checkTab(e) {
     updateTabbableNodes();
-    if (e.shiftKey && e.target === state.firstTabbableNode) {
-      e.preventDefault();
-      tryFocus(state.lastTabbableNode);
-      return;
+    const isShift = e.shiftKey;
+
+    if (isShift) {
+      const startOfGroupIndex = state.tabbableGroups.findIndex(
+        ({ firstTabbableNode }) => e.target === firstTabbableNode
+      );
+
+      const isStartOfAGroup = startOfGroupIndex !== -1;
+      if (isStartOfAGroup) {
+        e.preventDefault();
+
+        const isFirstGroup = startOfGroupIndex === 0;
+        const lastGroupIndex = state.tabbableGroups.length - 1;
+        const previousGroupIndex = startOfGroupIndex - 1;
+        const destinationGroupIndex = isFirstGroup
+          ? lastGroupIndex
+          : previousGroupIndex;
+
+        const destinationGroup = state.tabbableGroups[destinationGroupIndex];
+        const destinationNode =
+          destinationGroup && destinationGroup.lastTabbableNode;
+
+        tryFocus(destinationNode);
+
+        return;
+      }
     }
-    if (!e.shiftKey && e.target === state.lastTabbableNode) {
-      e.preventDefault();
-      tryFocus(state.firstTabbableNode);
-      return;
+    if (!isShift) {
+      const lastOfGroupIndex = state.tabbableGroups.findIndex(
+        ({ lastTabbableNode }) => e.target === lastTabbableNode
+      );
+
+      const isLastOfAGroup = lastOfGroupIndex !== -1;
+      if (isLastOfAGroup) {
+        e.preventDefault();
+
+        const isLastGroup =
+          lastOfGroupIndex === state.tabbableGroups.length - 1;
+        const firstGroupIndex = 0;
+        const nextGroupIndex = lastOfGroupIndex + 1;
+        const destinationGroupIndex = isLastGroup
+          ? firstGroupIndex
+          : nextGroupIndex;
+
+        const destinationGroup = state.tabbableGroups[destinationGroupIndex];
+        const destinationNode =
+          destinationGroup && destinationGroup.firstTabbableNode;
```

Comment:
Similarly to the shift case, I think we can simplify this down to 3 statements and 3 variables.

In fact, we could also de-dup the `tryFocus(destinationNode)` statement by declaring `let destinationNode` before the IF statement, and then call `tryFocus(destinationNode)` after the IF statement, as the last thing the function does before returning.

Reply:
Done